packages/browser/src/node/commands/screenshotMatcher/index.ts

```diff
@@ -0,0 +1,246 @@
+import type { BrowserCommand, BrowserCommandContext } from 'vitest/node'
+import type { ScreenshotMatcherOptions } from '../../../../context'
+import type { ScreenshotMatcherArguments, ScreenshotMatcherOutput } from '../../../shared/screenshotMatcher/types'
+import type { AnyCodec } from './codecs'
+import type { AnyComparator } from './comparators'
+import type { TypedArray } from './types'
+import { mkdir, readFile, writeFile } from 'node:fs/promises'
+import { dirname } from 'pathe'
+import { asyncTimeout, resolveOptions, takeDecodedScreenshot } from './utils'
+
+export const screenshotMatcher: BrowserCommand<
+  ScreenshotMatcherArguments
+> = async (context, name, testName, options): ScreenshotMatcherOutput => {
+  if (!context.testPath) {
+    throw new Error(`Cannot compare screenshots without a test path`)
+  }
+
+  const { element } = options
+
+  const {
+    codec,
+    comparator,
+    paths,
+    resolvedOptions: { comparatorOptions, screenshotOptions, timeout },
+  } = resolveOptions({ context, name, testName, options })
+
+  const referenceFile = await readFile(paths.reference).catch(() => null)
+  const reference = referenceFile && await codec.decode(await readFile(paths.reference), {})
+
+  const abortController = new AbortController()
+  const stableScreenshot = getStableScreenshots({
+    codec,
+    comparator,
+    comparatorOptions,
+    context,
+    element,
+    name,
+    reference,
+    screenshotOptions,
+    signal: abortController.signal,
+  })
+
+  const value = await (
+    timeout === 0
+      ? stableScreenshot
+      : Promise.race([
+          stableScreenshot,
+          asyncTimeout(timeout).finally(() => { abortController.abort() }),
+        ])
+  )
+
+  // case #01
+  //  - impossible to get a stable screenshot to compare against
+  //  - fail
+  if (value === null || value.actual === null) {
+    return {
+      pass: false,
+      reference: referenceFile && paths.reference,
+      actual: null,
+      diff: null,
+      message: `Could not capture a stable screenshot within ${timeout}ms.`,
+    }
+  }
+
+  const { updateSnapshot } = context.project.serializedConfig.snapshotOptions
+
+  // if there's no reference or if we want to update snapshots, we have to finish the comparison early
+  if (reference === null || updateSnapshot === 'all') {
+    const shouldCreateReference = updateSnapshot !== 'none'
```

Comment:
Another question from the note: should this check if ref and actual match and eventually set the flag to false? Would it even make sense to update screenshots if they still match?